src/widgets/panel/panel.js

```diff
@@ -129,13 +129,23 @@ export default function panel({
         container: getContainerNode(bodyRef),
       });
 
+      // All widget's methods need to be bound to not lose the `this`
+      // context when called from the `panel` widget.
+      // We need to check the type of the widget's attributes before binding them
+      // because some are not methods.
+      Object.keys(widget)
+        .filter(attribute => typeof widget[attribute] === 'function')
+        .forEach(method => {
+          widget[method] = widget[method].bind(widget);
+        });
```

Comment:
I'm not completely sure it will work in all cases. Because sometimes we create the function on the instance at the `init` step not ahead of time. If I'm not mistaken this code happens before the `init` lifecycle of the connector (like inside [`connectMenu`](https://github.com/algolia/instantsearch.js/blob/develop/src/connectors/menu/connectMenu.js#L180-L185)).

Reply:
That might be true.
We would either need to refactor those, or dump `this` from the connectors (initiative taken in #3347).

What other solution do you see?

Reply:
Actually it does work because it doesn't matter once we have bind the instance to the correct context the one inside `init` is also correct and so on. But I would avoid to bind all the function on the widget. What we can do is to re-use the instance we have to call the appropriate method on the widget. The instance we have in the `panel` is the one that is used so we can re-use it (see below).

```js
return {
  // ...
  dispose(...args) {
    if (typeof widget.dispose === "function") {
      widget.dispose.call(this, ...args);
    }
  }
};
```

Reply:
We sometimes rely on `this` in [non-conventional methods](https://github.com/algolia/instantsearch.js/blob/develop/src/connectors/menu/connectMenu.js#L135-L140) in the connectors. This would break the implementation if we don't bind these methods.

What's wrong with binding all the methods for now?

Reply:
I agree, but the solution with `call`, or binding _only_ the public APIs, will fail in some cases (e.g. [`connectMenu`](https://github.com/algolia/instantsearch.js/blob/develop/src/connectors/menu/connectMenu.js#L135-L140)).

Reply:
Did you tried the version with `call` and `panel` over `menu` ? I don't think it fails.

Reply:
Indeed, it works!